### PR TITLE
Add experimental deterministic flag

### PR DESCRIPTION
**Pull request by kurtamohler**
Adds `torch.experimental.deterministic` flag to enforce deterministic algorithms across all of pytorch.
Adds `torch.experimental.deterministic_error_level` to allow users to choose between error/warning/silent if determinism for an operation is not available.
Adds `torch.experimental.alert_not_deterministic()` which should be called within operations that are not deterministic.
Offers both Python and ATen interfaces

Issue #15359

### CODE REVIEW

**Comment by kurtamohler** [comment]
In addition to creating the deterministic flag API, I've also made these changes so far:
* Make bmm sparse-dense CUDA use the flag
* Make `at::convolution` use the flag
* Call `alert_not_deterministic()` in `torch.nn.functional.interpolate`

I'll need to find all the appropriate operations in pytorch that should use the flag or alert. Does there happen to be a list of these functions, or any advice on finding them myself?

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -688,7 +688,7 @@
     SparseCUDA: bmm_sparse_cuda
   supports_named_tensor: True
 
-- func: _bmm(Tensor self, Tensor mat2, *, bool deterministic=False) -> Tensor
+- func: _bmm(Tensor self, Tensor mat2, *, bool? deterministic=None) -> Tensor
```
Comment (kurtamohler) [inline aten/src/ATen/native/native_functions.yaml:691]: Note that I made this change so that if the deterministic argument is not included in a `_bmm()` call, it will use `torch.experimental.deterministic`, rather than `False`

**Comment by kurtamohler** [comment]
Because of the changes I made to `_bmm`, it's getting flagged as bc-breaking. But since I'm just changing from a bool kwarg to an optional bool kwarg, there won't really be any changes in how callers interact with it. Does this truly qualify as a bc-breaking change?

**Comment by ezyang** [comment]
cc @houseroad

**Comment by kurtamohler** [comment]
Since the amount of functions that could use the deterministic flag is potentially huge, I wonder if we should address them in future PRs.

**Comment by kurtamohler** [comment]
Ralf suggested that I make a wiki page about how to use the deterministic flag API so that other people who know which operations have determinism/nondeterminism can add the appropriate calls. We also considered that there might be large groups of operations that are implemented on top of some shared API which could potentially be modified to use the deterministic flag underneath, rather than explicitly adding the deterministic checks to each individual operation.

**Comment by ezyang** [comment]
Sorry about this PR falling off the wagon.

> I'll need to find all the appropriate operations in pytorch that should use the flag or alert. Does there happen to be a list of these functions, or any advice on finding them myself?

There isn't going to be a list. People like @ngimel or @colesbury may be able to tell you about the folklore known to be nondeterministic operators, and also have ideas about what types of operators are likely to nondeterministic, but getting the full list is going to take work, and you're probably going to want a plan going into that work. Pursuant to that...

> Since the amount of functions that could use the deterministic flag is potentially huge, I wonder if we should address them in future PRs.

Yes, I agree. Start with stuff we know about, gradually make it better over time. A wiki page that says how to figure out if things are deterministic or not would be super helpful.

Also the deterministic argument to a function is orthogonal to just warning when things are nondeterministic, and we can do that without worrying about BC implications.

This patch is still labeled WIP; are you looking for reviews now?

**Comment by t-vi** [comment]
`atomicAdd` is the main culprit and grepping for it found most of the things in the nondeterminism note.

**Comment by kurtamohler** [comment]
Removed WIP label. Yes, I think it's at a good point for review.

File: aten/src/ATen/Context.cpp [hunk 2]
```diff
@@ -59,6 +59,48 @@ void Context::setDeterministicCuDNN(bool b) {
...
+
+Context::ErrorLevel Context::deterministicErrorLevel() const {
+  return deterministic_error_level;
+}
+
+void Context::setDeterministicErrorLevel(Context::ErrorLevel e) {
+  deterministic_error_level = e;
+}
+
+Context::ErrorLevel Context::longToErrorLevel(long e) {
```
Comment (ezyang) [inline aten/src/ATen/Context.cpp:78]: Not sure why this is a method on Context, it has nothing to do with the Context...

File: aten/src/ATen/Context.cpp [hunk 3]
```diff
@@ -59,6 +59,48 @@ void Context::setDeterministicCuDNN(bool b) {
...
+      TORCH_CHECK(false, "error level ", e, " is invalid, must be one of 0: None, 1: Warn, or 2: Error");
+  }
+}
+
+void Context::alertNotDeterministic(std::string const& caller) {
+  if (globalContext().deterministic()) {
+    auto error_level = globalContext().deterministicErrorLevel();
+    switch (error_level) {
+      case Context::ErrorLevel::Warn:
+        TORCH_WARN(caller, " is not deterministic");
```
Comment (ezyang) [inline aten/src/ATen/Context.cpp:94]: It would be better to have a more detailed error message explaining why this error message is happening (it's because you set the deterministic flag) and what the user should do in this situation (whether it's file a bug asking for deterministic support, or turn off the flag, or something else.)

File: torch/experimental/__init__.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import sys
```
Comment (ezyang) [inline torch/experimental/__init__.py:1]: Don't forget to add documentation for this module

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu [hunk 5]
```diff
@@ -895,6 +895,7 @@ Tensor& _bmm_out_sparse_cuda(Tensor& result, const SparseTensor& self, const Ten
   size_t workspace_buffer_size = 0;
   void* workspace_buffer = nullptr;
 
+  deterministic |= globalContext().deterministic();
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:898]: `deterministic = deterministic || globalContext().deterministic()` is slightly better because you get to short circuit if `deterministic` is true already.

File: aten/src/ATen/Context.h [hunk 6]
```diff
@@ -94,6 +94,8 @@ class CAFFE2_API Context {
 
   bool setFlushDenormal(bool on);
 
+  enum ErrorLevel { None=0, Warn, Error };
```
Comment (ezyang) [inline aten/src/ATen/Context.h:97]: I think I would recommend (1) making this an enum class, (2) naming it more descriptively so that it is specifically about determinism options, (3) moving it outside of Context class

File: aten/src/ATen/Context.cpp [hunk 7]
```diff
@@ -59,6 +59,48 @@ void Context::setDeterministicCuDNN(bool b) {
...
+    case static_cast<int>(Context::ErrorLevel::None):
+    case static_cast<int>(Context::ErrorLevel::Warn):
+    case static_cast<int>(Context::ErrorLevel::Error):
+      return static_cast<Context::ErrorLevel>(e);
+    default:
+      TORCH_CHECK(false, "error level ", e, " is invalid, must be one of 0: None, 1: Warn, or 2: Error");
+  }
+}
+
+void Context::alertNotDeterministic(std::string const& caller) {
```
Comment (ezyang) [inline aten/src/ATen/Context.cpp:89]: nit: There isn't really any reason to take a const reference to a string here; if the user doesn't have a `std::string` hand all you're doing is forcing them to allocate one, just so that you can read it out. Pass a `const char*` or maybe a `c10::string_view` (if you're feeling fancy)

File: test/test_torch.py [hunk 8]
```diff
@@ -115,6 +115,39 @@ def make_contiguous_slice(size, dtype) -> torch.Tensor:
...
+                for error_level in [0, 1, 2]:
+                    with torch.experimental.flags(deterministic=deterministic, deterministic_error_level=error_level):
+                        self.assertEqual(deterministic, torch.experimental.deterministic)
+                        self.assertEqual(error_level, torch.experimental.deterministic_error_level)
+
+            with torch.experimental.flags(deterministic=True, deterministic_error_level=0):
+                # Should not raise or warn
+                torch.experimental.alert_not_deterministic("caller")
+
+            with torch.experimental.flags(deterministic=True, deterministic_error_level=1):
```
Comment (ezyang) [inline test/test_torch.py:129]: Magic numbers for the error level doesn't seem like very good API.  More idiomatic Python API is to use some sort of string like "warn" or "error" to switch between these cases.

**Comment by ezyang** [comment]
This seems like it's most of the way there. I'd really like it if @colesbury could take a look at the API and see what he thinks. I'm not a big fan of using numeric codes to control error level and suggest we should do something string based. Maybe before you code you could propose a full user API here first.

As a broader question, we should think about how we can actually get ahead of the curve in identifying when operations are deterministic or not. In general, the average person writing a new operator is not going to know about this API. If we have some op addition checklist, checking if it should be deterministic should be added to it; but we don't have such a checklist right now. Other possibilities are lint rules for things like atomicAdd, which have been mentioned as one big way we find nondeterminism.

**Comment by colesbury** [comment]
1) I don't think the configurable warning/error level is worth the API complexity. Just choose one: error or warning.

2) Similarly, I don't think the API complexity of a `flags` function is justified; this should just be a settable attribute `deterministic`. (Or a function `set_deterministic(True)` if there is good reason to prefer that)

3) Are there any uses of `alert_not_deterministic` in Python? I'd expect all the uses to be in C++. (Not a big deal either way)

4) Finally, I think `deterministic` just be on the `torch` package. I don't think the `experimental` sub-package buys us anything; it doesn't make API breakages any less frustrating for users and we'll either have to move it to `torch` anyways or it'll stay in experimental forever. We've buried things in generically named sub-packages before (`torch.backends`) and it's only led to confusion.

So what I'd like to see as the public Python API is either:

```
torch.deterministic = True
```

or

```
torch.set_deterministic(True)
torch.is_deterministic() -> bool
```

and on the C++ side:

```
torch::set_deterministic(true)
bool torch::is_deterministic()
```

**Comment by ezyang** [comment]
```
torch.deterministic = True
```

will probably be hard to implement, so we should go for the function api instead.

**Comment by kurtamohler** [comment]
Thanks for all the input!

**Comment by kurtamohler** [comment]
If we're getting rid of the error level setting, would it be better to error or warn? I feel like an error would be better. If a user explicitly asks for a feature that isn't available, it seems like we should make it impossible for them not to notice that it's unavailable. If we just print a warning, then the user could potentially miss it, especially if their application prints a lot of other messages or warnings. And the user could easily get around the error by disabling the deterministic flag only for the offending operation.

**Comment by t-vi** [comment]
Advantages of warning:
- you can still use PyTorch,
- you can actually choose to promote warnings to errors in Python.
Disadvantages of warning:
- stack trace?

I must admit I see little wrong with keeping a state with three possibilities because all the difference in handling will be contained in a single function but hey, what do I know.

**Comment by colesbury** [comment]
> If we're getting rid of the error level setting, would it be better to error or warn? 

If you're unsure, go with an error; most people seemed to prefer that option.

**Comment by kurtamohler** [comment]
@colesbury , I have a question about your comment here:
> I don't think the API complexity of a flags function is justified

I feel like the convenience of being able to use a context manager for the deterministic flag is worth the extra API complexity. A few sections of an application need to always be deterministic/non-deterministic regardless of the global setting.

For instance, imagine if someone is using one operation somewhere that doesn't have a deterministic implementation. Given that we're deciding to error out when a deterministic implementation is unavailable, then the user would need to do something like this:

```
torch.set_deterministic(true_or_false)
...
restore_deterministic = torch.is_deterministic()
torch.set_deterministic(False)
torch.some_nondeterministic_op()
torch.set_deterministic(restore_deterministic)
...
```
But if we allow the user to change the setting through the flags function, they would be able to do this:
```
torch.set_deterministic(true_or_false)
...
with torch.flags(deterministic=False):
    torch.some_nondeterministic_op()
...
```
In my opinion, the context manager makes it much more readable and less error-prone.

**Comment by colesbury** [comment]
@kurtamohler a context manager introduces it's own problems: if the state is truly global then changing it is not thread-safe. If it's thread-local, then it's not inherited by new threads, which is confusing. We've run into this issue with `with torch.set_grad_enabled(...):`. (And to be clear, I think a true global variable is a better choice for torch.set_deterministic.) Context managers look like they have lexical scoping, but in this case the effect would be global. 

Additionally, I'm not sure that selectively enabling determinism will come up often (or even at all). A non-deterministic sub-component generally makes the entire computation non-deterministic. One can imagine possible uses, but we don't have any real examples yet. Finally, if needed, we can add support for context managers in the future using the syntax you mention in your edit without breaking backwards compatibility.

**Comment by kurtamohler** [comment]
Cool, makes sense! Thanks @colesbury

**Review by ezyang — APPROVED** [review]
Thanks, this is short and sweet.

**Comment by kurtamohler** [comment]
I don't think the `caffe2-py3.6-devtoolset7-rocmrpm-centos7-test` failure is my fault, though I'm having trouble figuring out where the actual failure is in the log. But it looks like this CI job is failing with a similar signature in other recent PRs.

**Comment by kurtamohler** [comment]
Looks like the internal checks failed :/
Anything I can do to fix it?

**Comment by ezyang** [comment]
So far, they're fake. I'll let you know if you need to do anything

**Comment by zou3519** [comment]
@kurtamohler this change made it into the 1.6 release. However, the docstrings for `torch.set_deterministic` imply that the feature is complete, i.e., if an operator is not deterministic, then using `torch.set_deterministic` will definitely throw a RuntimeError. 

> Sets a global flag to force all operations to use a deterministic
implementation if available. If an operation that does not have a
deterministic implementation is called while this setting is True, the
operation will throw a RuntimeError.

According to [this wiki page](https://github.com/pytorch/pytorch/wiki/How-to-support-%60torch.set_deterministic()%60-in-PyTorch-operators) it looks like indeed the feature isn't complete.

Because of that, we'd like to change `torch.set_deterministic` and `torch.is_deterministic` to `torch._set_deterministic` / `torch._is_deterministic` in the 1.6 release branch and/or add a warning when a user attempts to use these. Does that sound reasonable?
